### PR TITLE
[FLINK-28052][tests] Remove RunFailedJobListener 

### PR DESCRIPTION
**Pull request by zentol**
The RunFailedJobListener had rather obscure semantics.
It considered a job to be terminal after it was restarted. This is awfully specific to a particular test case.
A cleaner approach is just to just cancel the job and wait for it to terminate.

Additionally it considered a job as running purely based on the job status, whereas, in particular when checkpointing is involved, waiting for the tasks to be submitted is a better measure.
In fact, testExceptionHistoryWithTaskFailureFromStopWithSavepoint was a broken since a savepoint was never triggered, as not all tasks were running.

This PR also contains a few cleanup commits.

### CODE REVIEW

**Comment by zentol** [comment]
> But I don't understand why it's important to know that one test issues task restarts for the old implementation to be kind of reasonable?

It doesn't make sense because it's not even doing what it says on the tin.
What it looks like the test is doing is completing some job cancellation (that came from _somewhere_) and waiting for it to reach a terminal state.

In case of the restarting test, it just completes the restart (==acknowledging the cancel requests). The job actually keeps running. Which is then covered by this special case in the listener where we consider it terminal if it restarted.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java [hunk 1]
```diff
@@ -1080,10 +1053,10 @@ private Iterable<RootExceptionHistoryEntry> runExceptionHistoryTests(
                             declarativeSlotPool,
                             createSlotOffersForResourceRequirements(
                                     ResourceCounter.withResource(
-                                            ResourceProfile.UNKNOWN, numAvailableSlots)),
+                                            ResourceProfile.UNKNOWN, PARALLELISM)),
                             taskManagerGateway);
                 });
-        listener.waitForRunning();
+        taskManagerGateway.waitForSubmissions(4, TestingUtils.infiniteDuration());
```
Comment (XComp) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:1059]:
```suggestion
        taskManagerGateway.waitForSubmissions(PARALLELISM, TestingUtils.infiniteDuration());
```
We can remove the magic number here as well...

Comment (XComp) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:1059]: can we add a comment here about why waiting for all subtasks here is necessary due to checkpointing?

**Review by XComp — APPROVED** [review]
Ah, I see. I missed the if condition in the `JobStatusListener` implementation. Now, it makes sense. Looks good to me 👍 

nit: There's a "[...] A cleaner approach is just to just cancel [...]" in the commit message of the last commit that you could fix...

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java (hunk 1 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java]: good point; will do

**Pull request merged**